### ISSUE TITLE
Default container names specific to a browser window

### ISSUE THREAD

**Issue by Dannyzen** [mozilla/testpilot-containers#319]
Hi there,
I'm a big fan of the profiles functionality in firefox. Containers look support the use case I'm looking for in a modern way. My use case is pretty simple: I have multiple accounts on account specific sites (azure, aws, google, etc.) I use these services for both work and home. I use different browser windows to hold work accounts and home accounts. 

This feature request is to be able to have Firefox open a specific container (as its default) in a browser window so that my sessions can be segmented and support a "this window has home tabs" and "this window has work tabs" workflow

**Comment by lmorchard** [comment]
Yeah, this is my workflow too. Making tab-by-tab container decisions is a lot of mental overhead, IMO.

I basically segment my contexts with windows. Or a window in another browser if I need to separate logins (e.g. on Google). Or a private window if I want a disposable context. I'd also love to be able to naturally move tabs between two different windows to switch containers. And when I hide a container, I'd want restoring it to try to restore the original window arrangement - not pile all the tabs into one window.

**Comment by Teyras** [comment]
This would help my workflow a lot! It would also be great if the container name could be added to the window title - this way, I could handle each container differently in my WM.

**Comment by timeassistant** [comment]
I would like to add an infinite amount of +1s to this ! 

I'm a user of Multifox (or was) with the "window mode" option enabled , as its no longer supported I can no longer assign a profile to an entire window, my workflow is similar to the previous posters. 
![image](https://user-images.githubusercontent.com/18701905/31026513-ae25e52e-a53e-11e7-8626-a90a60ff747c.png)

Setting profiles on a per tab basis does not work for me as I can have anywhere upto and over 170odd tabs across several windows. 

Being able to set a profile per window (or even option to have it per window or per tab to appease everyone) would be a massive advantage and would definitely help to streamline a very busy work flow !

**Comment by jonathanKingston** [comment]
To implement this we would likely need a platform change to do this cleanly. Other extensions have been able to implement this however I would rather not make hacks.

See: https://bugzilla.mozilla.org/show_bug.cgi?id=1403422

**Comment by Dannyzen** [comment]
Thanks @jonathanKingston for the context. Is there a higher likelihood that we might expect this feature from the core mozilla team or an extension (assuming someone builds it! 😆 ) in the near future?

**Comment by jonathanKingston** [comment]
Not sure, "container windows" is certainly well asked for and this is essentially that somewhat. I think we should keep pushing for the platform support for this feature first. Even if we don't end up making it an option it will allow addons to make this much more cleanly.

**Comment by groovecoder** [comment]
Note: this is also a commonly-mentioned feature for Chrome-like parity to help users switch from Chrome. ;)

**Comment by timeassistant** [comment]
In addition to the note from @groovecoder - it will also help to stop users switching over to chrome now that there is currently no easy way to set container windows with Firefox anymore, it may not seem like much but it is a deal breaker for many users

**Comment by Dannyzen** [comment]
If I'm reading @jonathanKingston right, there's a higher likelihood the extendability will be built out _before_ this becomes a first class feature, resulting in an opportunity for an extension to add the feature. 

If this is the case, I'd be happy to put a bounty on the extension and ask if the others on this issue would be interested in joining me.

**Comment by sandinmyjoints** [comment]
This and bookmarks are the only things keeping me from switching entirely to Firefox -- hope it happens soon!

**Comment by gustafbstromuu** [comment]
I would definitely _love_ if I could set a default container – this is would enhance the Firefox browsing experience by some order of magnitude!

**Comment by DHager** [comment]
@sandinmyjoints I too want both of those things (bookmarks and window-defaults) but we should probably consider the scenario where they *conflict*.

In particular, when someone uses a Container-X-by-default window, and selects a Container-Y bookmark.

Personally, I think the bookmark's settings should win, it's the "privacy safest" outcome. If someone's workflow is to use the same bookmark in multiple identities, then it's probably best for them to *not* set a specific container for that bookmark at all.

**Comment by chicagobuss** [comment]
Happy new year, all.  Any news on this effort?  Really liking containers but wish I could configure a default for windows (and a default container overall) - when I click links in slack for work it's very annoying that they don't open in the work container!!

**Comment by DHager** [comment]
OK, so it sounds like we're up to four related mechanisms:

1. Domain-based (in Firefox 57)
2. Bookmark-based (in Firefox 59)
3. Window-based (?)
4. Global default (?)

@chicagobuss While it would be great to set a "Work" container a the default for external links when on an actual at-the-office desktop, I have a harder time imagining the "right" behavior for a multipurpose machine. Firefox can't know whether the link you clicked was from a work-conversation versus a personal-one.

**Comment by sandinmyjoints** [comment]
@DHager re: default-container windows and bookmarks colliding: I'd prefer to see container-bound windows, in which it is not possible to open a tab that uses a different container. One (or more) window(s) for Work containers only, others for Personal, etc. Then bookmark conflicts go away because you can only see/navigate to bookmarks associated with container bound to the currently active window. 

This is what I'm used to in Chrome, so I'm very comfortable with this and my bookmarks are organized according to this principle. I can imagine use cases where you want to have tabs from different containers next to each other in the same window, like comparing a work account with a personal account for some reason. I'm sure there are creative ways to accommodate that use case.

**Comment by joesanford** [comment]
What about, instead of setting the context of a window, the functionality is mirrored in another way? For instance, any opened tab that originated from a tab within a container opens in the same container? Is this more feasible in the framework?

**Comment by soar** [comment]
Window-based behavior should be great enhancement! Will wait for it!

**Comment by cgchoffman** [comment]
+1.  I'm constantly opening "generic" tabs in the window I've designated for [personal | work | other] account.  I've only been using this tool for about 20 minutes and I already reeeeeally really want this :)

**Comment by DHager** [comment]
@sandinmyjoints @joesanford  What if instead of some mechanism for *explicitly* assigning a Window-Container relationship, it becomes *implicit* based on the population of tabs already present inside that window?

Suppose the user creates a tab with `ctrl-t`, which always happens in the context of a particular window. Check if that window *exclusively* contains tabs bound to a single container X. If so, then it is safe to assume that the user wants another tab for the container X as well. If not, then fall back to the current behavior.

In other words, you create a "Work Window" simply by making sure that none of the tabs in it are anything except Work tabs, and after that Firefox helps you avoid accidentally breaking that pattern.

**Comment by digitalowlnyc** [comment]
@DHager @sandinmyjoints @joesanford 

> What if instead of some mechanism for explicitly assigning a Window-Container relationship, it becomes implicit based on the population of tabs already present inside that window?

Interesting proposal, it seems like there are at least three cases here: 
1) when a user clicks a link outside of the browser with no context available
2) when a user clicks a link inside an existing tab. 
3) when a user clicks a bookmark

For case #1, what about simplifying and just using the last focused container (or window)? I personally think any implicit workflows should be as simple as possible for a user to easily reason about what will happen. This solution goes along with user workflow IMO, because typically I am already looking at a "work" tab whenever I am clicking links in other programs related to work. I can live with the compromise of accidentally clicking a link meant for personal and having it open in my work tab, or vice versa. This solution can potentially be applied to #3 as well.

For case #2, it seems like what makes sense is to stick to the same container for any clicked links.

In general, I like the "one container per window" approach as it seems like it would simplify user workflows.

I have been a Chrome user the past 4-5 years and would love to make the switch back to Firefox - this is the one thing holding me back. I am a developer and am open to being involved with development on this (in FF or a plugin) as a way to pitch in and give back to the FF community.

**Comment by cgchoffman** [comment]
+1 on @digitalowlnyc comment.   The 3 contexts and how they are handled as well as helping out.  I'm a Dev as well.

I don't think windows should be implicitly assigned an accnt though.  A window should be opened under a specific accnt.  Any tab opened in that window opens with that accounts settings.  The window has it's assigned colour and it's icon so I know which window is which. 

Again, if it's feasible to parcel out tasks I'd love to help out.  I guess I could just look at the source.

**Comment by DHager** [comment]
@digitalowlnyc  

> Interesting proposal, it seems like there are at least three cases here:

Adding two more:

* Case 4:  User creates an URL-less tab in an existing window (e.g. `ctrl-t`) and wants it to already be for a certain container.
* Case 5:  User creates an URL-less tab in a fresh window (e.g. `ctrl-n`) and wants it to already be for a certain container.

I find Case 4 is especially common when the next thing I'm about to do is type in some web-search terms.

**Comment by cgchoffman** [comment]
> Case 4: User creates an URL-less tab in an existing window (e.g. ctrl-t) and wants it to already be for a certain container.

This was the use case that made me go back to Chrome.  I was constantly opening a "session less" tab.

**Comment by Rukenshia** [comment]
I'd really love this. Another use case for me is that when I'm at work and need to log in to a CLI (`fly`, we are using [concourse](https://concourse.ci) and you log in via oauth), I have to click a link in my terminal. In my browser, I'm only logged in to the application in my "Work" tabs and by default when I Ctrl+Click the link in my terminal it will open a default tab, hence failing the login.

**Comment by ericholscher** [comment]
Trying to switch full-time from Chrome, and this is really holding me back. I really want to have separate browser windows for contexts, and have each of them default to different containers.

Is there an easy way to enable this in an extension or other format?

**Comment by DHager** [comment]
@ericholscher I can't speak to the technical issues, but I think the immediate problem is what workflows should -- and shouldn't -- be supported, or else how to get them to play along in a scheme that is understandable by users.

**Comment by kemayo** [comment]
I'm also in the camp of people who use this workflow in Chrome, and have it as a fairly major blocker for switching.

In my case, it's even more extreme separation, as I have multiple desktops and keep solely Work windows on one and solely Personal windows on another. This does completely remove any confusion over what should happen with external applications -- if I'm on the work desktop, all the other apps are also work-related, so of course I want links from them to open in my work profile.

There was an earlier statement of how this could be a good extra addon, but it needed some backend changes before that was practical. I'd be happy to poke at making that addon, if anyone knows whether those changes have happened?

**Comment by skinner** [comment]
The chrome design (tying the window to the container the way @sandinmyjoints describes it) is the right one.  I wanted the same behavior in firefox, so I ended up running multiple profiles concurrently by starting up firefox with separate `firefox -no-remote -P <profile name>` commands on the command line.

The experience ends up being very much like chrome in that identity, bookmarks, history, etc. are separate per-profile.  The chrome UX for creating those windows is way better, with the button in the top-right corner of the window, but otherwise the behavior is very similar.

I still think mozilla should implement some version of this issue to streamline the UX for this usage pattern.  But for people who want something now, profiles are there.

**Comment by keehun** [comment]
Has there any progress on this front? I agree that this is a really important next step for Multi-Account Containers

**Comment by DHager** [comment]
@keehun Frankly I don't think there's any consensus on basic behavior, especially because "do it like Chrome" would actually *removes* features/workflow that people are currently using in Firefox.

**Comment by Pablohn26** [comment]
This is the third most commented and most liked issue.

And I guess I am not the only one person here this issue is the main reason (and maybe the only reason) why we are still using chromium instead of Firefox.

This could be addressed as this issue explains or as #462 issue (the issue with most reactions) explains, but A LOT of people (the most of them do not care to arrive here) needs a way to keep two sesions/profiles/container in a quicker way.

**Comment by ivanhercaz** [comment]
I am very interested in this issue. Containers are an awesome feature, but they need to be refined. One of the things to improve is what @Dannyzen and many other people have suggested in this issue.

But the issue was opened in 2017, and now, one year and more later, after read this issue, I didn't read something about this suggested enhancement. Anyone involved with the development of multi-account-containers know if there has been any progress or if this is bogged down?

**Comment by tom-james-watson** [comment]
You can achieve this by using the Sticky Containers extension - https://addons.mozilla.org/en-US/firefox/addon/sticky-containers/reviews/.

**Comment by chicagobuss** [comment]
Really appreciate the note, @tom-james-watson - this basically achieves the workflow I was hoping for.  External links from slack just use the currently-open container for context and that's PERFECT as far as I'm concerned.

**Comment by byronmansfield** [comment]
I just wanted to express interest on this feature to keep this thread going. Also following @tom-james-watson comment about the extension, that is great someone made one to add this feature. Though the recent comments seem to mention it's not working for them. I'm guessing it's not actively maintained? I might give it a shot later. And also agree it should be either a default behavior or a built in feature.

**Comment by tom-james-watson** [comment]
Yeh I'd also point out that I ended up stopping using that Sticky Containers because it had a few annoying bugs. If I remember it made reopening of closed tabs work in strange ways. Would be far better to have this built into the browser itself.

**Comment by grnassar** [comment]
Also having some issues with Sticky Containers, and hoping that that functionality finds its way here -- and even better if (as in .@tom-james-watson 's comment 4 above this one) either made the default or could be configured to be default in options.

**Comment by andrewleech** [comment]
Not as comprehensive as Sticky Containers was, but https://addons.mozilla.org/en-US/firefox/addon/sticky-window-containers/ does a similar thing. 
Whatever container the first tab it in, any new tabs open in the same container.

**Comment by tom-james-watson** [comment]
That actually works far better than Sticky Containers - thanks!

**Comment by tom-james-watson** [comment]
That actually works far better than Sticky Containers - thanks!

It's a shame there doesn't seem to be any first-party activity on addressing this.

**Comment by Snogren** [comment]
Thanks @andrewleech! Works for me.

**Comment by adamshand** [comment]
I also really want this feature for the same reason.  I'm a freelancer so sometimes have windows open for half a dozen different clients all logged into similar sites.  Containers are great but it's annoying to have to constantly right click "reopen tabs in container" and it's easy to make mistakes and end up logged in as the wrong user.  I also manage this by having a window per client.

Thanks for the recommendation for sticky window containers, that a big improvement!

**Comment by grnassar** [comment]
> Not as comprehensive as Sticky Containers was, but https://addons.mozilla.org/en-US/firefox/addon/sticky-window-containers/ does a similar thing.
> Whatever container the first tab it in, any new tabs open in the same container.

That is a *huge* improvement over stock. Thank you!

**Comment by githubflub** [comment]
Hello???

**Comment by moontails** [comment]
happy to work on this :)

**Comment by michalfita** [comment]
I didn't care that much but I moved jobs where Google is now the productivity suite with everything in the browser. Keeping the stuff separated by the Window would help a lot. A function sending a tab opened in default work windows into a private window would help as well.

**Comment by michalfita** [comment]
I didn't care that much but I moved jobs where Google is now the productivity suite with everything in the browser. Keeping the stuff separated by the Window would help a lot. A function sending a tab opened in default work windows into a private window would help as well. Sticky Windows Containers tries to address this problem, but it's not good enough especially for dealing with where tabs are open from requests coming from outside the browser.

**Comment by nemothorx** [comment]
containers have become one of the killer features for me in FF, and I've spruiked it to several colleagues successfully. 

But I wanted more. I wanted different windows to have different default containers. This seems to be the thread to voice my support for this type of functionality!

I do notice a lot of people want windows to be limited to a single container - that wouldn't work for me. By and large I rely on being able to mix containers within a window. 

My ideal (so I imagine) case would simply to be able to configure a window as having a default container, which applies to new tabs when created interactively ([New tab] menu item, or equivalent hotkey)

I still want new tabs within other containers to be possible. I'd also want links opened from external applications to default to the default session.

**Comment by achernyakevich-sc** [comment]
I could advise to try Simple Tab Groups add-on. In my case it resolve my problem and works not only for window:

https://addons.mozilla.org/en-US/firefox/addon/simple-tab-groups/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search

In few words: you could define in setting that the Container `C-A` should be always open in the Group `G-A` and as soon as Group `G-A` is open in separate windows then new tab that should belong to `C-A` Container will open there.

In case of you wish to open new tab using the same container as current tab then I would advise to vote to introduce `Ctrl+Shift+0` shortcut in the https://github.com/mozilla/multi-account-containers/issues/462#issuecomment-899382995 :)

**Comment by r-0-0-t** [comment]
+1 for this feature

**Comment by grahamperrin** [comment]
Please use the upvote feature. (Not written +1)

**Comment by achernyakevich-sc** [comment]
> While I use and appreciate both simple tab groups and tab containers, neither truly provides a complete solution to the individual problems and they definitely don't work together an overall solution to make what everyone seems to be asking for simple and efficient.

I use both at the same time - MAC + STG. And it works perfect because STG knows about MAC existence and can be configured any way you would like it should work. But I agree that user should configured it as he need and there will be no one "silver bullet" configuration that will match to all possible usage scenarios. It is like philosophy law - if you have universal solution then you will pay something for this universality - either performance or efforts for sharp tuning and configuration. :)

**Comment by ejdiener** [comment]
I know I'm years late to the party, but want to again bring this up! It would be so helpful if I could have a default tab type set for each window.
